Fixes for ATT timeout

jori-nordic opened the pull request:
Fixes ATT timeout when stack is under memory pressure.

The gist of it is:
- The host stack (ATT in this case) does not involve the application when handling REQuests
- ATT _requires two resources_ to enqueue a RSP: a buffer and a metadata struct
- These are allocated from two shared pools: `acl_tx_pool` and `tx_meta_data` 
- The global buffer pool is shared by all users of the stack (l2cap coc, smp, att)
- The metadata pool is shared by all users of att (can be multiple threads sending notifications)

If the application is putting pressure on those buffer and metadata pools (e.g. by sending notifications very very fast to multiple devices), then ATT can be starved and will fail to enqueue a response. 
Then the ATT timer will fire and we will get an ATT timeout. 

That is pretty bad in the current version of the host, as - apart from a log message - the application is not notified of this. What it can do is try to use GATT (eg send a notification) and carefully inspect the error value to determine if the ATT bearer has been shut down.

This PR proposes a two-part fix:
- add a struct member for the att metadata per-channel
- re-use the REQ buffer to enqueue the RSP

This way those two necessary resources are always available, as there can only ever be one transaction at a time on a bearer/channel.

jori-nordic committed:
Bluetooth: att: return INSUFFICENT_RESOURCES on PDU alloc failure

Easier to debug that way.

Ideally we'd have more error codes/logging instead of just returning
UNLIKELY for most errors.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: att: use a dedicated metadata struct for RSP PDUs

The ATT module has provisions to queue a packet/buffer for sending later if
it can't send it right away. For example if the conn.c tx context
allocation fails.

This unfortunately doesn't work if the buffer can't get allocated in the
first place, or if the ATT metadata can't also be allocated.

The metadata is allocated from a global pool set to the same number as
conn.c TX contexts. That can lead to a situation where other users of ATT
manage to queue a bunch of buffers (e.g. the app spamming GATT
notifications), depleting the number of ATT metadata slots so that none are
available.

When none are available, and we receive an ATT REQ, we try to allocate one,
fail, and drop the buffer (!). That pretty much guarantees an ATT timeout.

As a workaround for this, use a per-channel metadata slot, that is only
used for completing transactions.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: att: log address on timeout

This is useful in multilink scenarios, especially since there is not user
callback when the ATT channel times out.

Adding a user-facing callback should ideally also be done, but just logging
the address already provides useful insight.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: att: re-use REQ buf for RSP

This to ensure we don't fail to send a response and never get an ATT
TIMEOUT due to ACL TX buffer starvation caused by other users of the stack.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -657,22 +667,47 @@ struct net_buf *bt_att_chan_create_pdu(struct bt_att_chan *chan, uint8_t op,
+
+		net_buf_reset(buf);
+		size_t reserve = sizeof(struct bt_l2cap_hdr);
+		reserve += sizeof(struct bt_hci_acl_hdr) + BT_BUF_RESERVE;
```
```suggestion
		reserve += BT_BUF_SIZE(sizeof(struct bt_hci_acl_hdr));
```

Thalley reviewed (commented):
A few minor comments. 

Since this is a recurring thing (clearly not caught by any of our tests), it seems that we are missing a test for this particular issue. Any chance you can implement such a test as well?

jori-nordic commented:
>  Any chance you can implement such a test as well?

Haha that's the problem actually! I did implement the test, but it's finding so many issues that we can't merge it anytime soon. It's a multilink stress test [see here for a draft](https://github.com/jori-nordic/zephyr/tree/e2d85c5b031bd9c4bd6e2882088db1513a744394/./), and I have to test w/ the softdevice because the ZLL crashes with scheduling asserts.

We're working with @alwa-nordic to remove all the blocking (for buffers, metadata contexts etc) in the host, and that test is helping us a lot. I'll try to merge it as soon as we have managed to do that.

Thalley commented:
If you want to verify that this is fixing some of the EATT issue we've seen, you can try to add 
```
CONFIG_BT_L2CAP_ECRED=y
CONFIG_BT_EATT=y
```

to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bsim/bluetooth/audio/prj.conf and see if it passes CI

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -91,6 +103,7 @@ struct bt_att_chan {
 	ATOMIC_DEFINE(flags, ATT_NUM_FLAGS);
 	struct bt_att_req	*req;
 	struct k_fifo		tx_queue;
+	struct bt_att_tx_meta_data transaction_meta;
```
rsp_meta

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -194,7 +195,11 @@ static inline void tx_meta_data_free(struct bt_att_tx_meta_data *data)
 
 	(void)memset(data, 0, sizeof(*data));
-	k_fifo_put(&free_att_tx_meta_data, data);
+
```
if (data == chan.rsp_data) { /* "Freeness" is kept track of by remote (you can only have one transaction). */ }
else {assert(part_of_array())}

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -657,6 +663,7 @@ struct net_buf *bt_att_chan_create_pdu(struct bt_att_chan *chan, uint8_t op,
 	case ATT_CONFIRMATION:
```
test cfm code path

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -668,11 +675,23 @@ struct net_buf *bt_att_chan_create_pdu(struct bt_att_chan *chan, uint8_t op,
+		 * module can then reschedule the RSP if it is not able to send
+		 * it immediately.
+		 */
+		__ASSERT_NO_MSG(chan->transaction_meta.att_chan == NULL);
```
Restore this invariant in the free().

jori-nordic replied:
it's already being restored (by the `memset`)

alwa-nordic replied:
As discussed offline: This can be intentionally triggered by an remote, so it should not be an assert, but rather a disconnect or other more graceful failure.

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -684,8 +681,27 @@ struct net_buf *bt_att_chan_create_pdu(struct bt_att_chan *chan, uint8_t op,
+		buf = net_buf_ref(chan->transaction_buf);
+
+		net_buf_reset(buf);
+		net_buf_reserve(buf, BT_L2CAP_BUF_SIZE(0));
```
Double check headroom requirement for EATT

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -2880,6 +2897,19 @@ static int bt_att_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+		 */
+		att_chan->transaction_buf = buf;
+	}
+
```
att_chan->transaction_buf = NULL;
before exiting this function, since it "borrows" from this function scope.

alwa-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3930,6 +3930,16 @@ static void rx_work_handler(struct k_work *work)
 }
 #endif /* !CONFIG_BT_RECV_BLOCKING */
 
+bool bt_is_wq(void)
```
:joy:

alwa-nordic reviewed (changes requested):
Internal review. Pay no mind if it makes no sense.

jori-nordic committed:
Bluetooth: att: allocate RSP buffer from same channel as REQ

Using a different channel for responding to a request is forbidden by spec.

The allocator was especially flawed as it iterated over all the EATT
channels to find one w/ a big enough MTU, but the sending was still done
over the same channel as the REQ.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: l2cap: add debug print for EATT

Add debug print when a buffer intended for ECRED is segmented
unnecessarily.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
@Thalley thanks for the audio tests heads-up, helped iron out some kinks (along with the EDTT tests).

Now there is only one test failing (also fails on main): `tests/bsim/bluetooth/audio/test_scripts/bap_unicast_audio.sh`

I attached the logs to this comment.
[audio-log.txt](https://github.com/zephyrproject-rtos/zephyr/files/12147393/audio-log.txt)

jori-nordic commented:
> Isn't the default actually CONFIG_BT_RECV_BLOCKING for most use cases?

Don't have usage statistics but all SoftDevice Controller users have the default `BT_RECV_WORKQ_BT`.

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -3251,6 +3343,8 @@ static struct bt_att_chan *att_chan_new(struct bt_att *att, atomic_val_t flags)
 	(void)memset(chan, 0, sizeof(*chan));
 	chan->chan.chan.ops = &ops;
 	k_fifo_init(&chan->tx_queue);
+	memset(&chan->rsp_buf, 0, sizeof(chan->rsp_buf));
```
`chan->rsp_buf` is a pointer, so it'd be clearer if you just did `chan->rsp_buf = NULL`. That said, setting a reference counted pointer to NULL without unreffing it first looks very suspicious (and the fact that you have a memset here potentially hides such a bug, since the reviewer doesn't necessarily understand that it's a pointer variable you're memsetting to zero).

jhedberg replied:
> Since it's the initial initialization, I can't really check `buf->ref` since there might be garbage in there.

Right, sorry, I missed that part. However, you have `(void)memset(chan, 0, sizeof(*chan));` just a few lines earlier, so it seems any additional zeroing is just redundant?

jori-nordic replied:
that's true, I can also drop the zeroing of the `meta` too

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -507,6 +525,8 @@ static void chan_rsp_sent(struct bt_conn *conn, void *user_data, int err)
 		atomic_clear_bit(chan->flags, ATT_PENDING_RSP);
 	}
 
+	chan->rsp_buf = NULL;
```
Missing `net_buf_unref(chan->rsp_buf)` before this. Please don't introduce this kind of implicit reference counting where you play around with reference counted pointers without using `ref()` and `unref(). It makes the code difficult to follow and review, and risks introducing bugs.

jori-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -2837,26 +2898,40 @@ static int bt_att_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
+	 * users (e.g. GATT notifications, L2CAP using global pool, SMP,
+	 * etc..), avoiding an ATT timeout due to resource usage.
+	 *
+	 * The ref is taken by `bt_att_chan_create_pdu`.
```
It does take a ref:
https://github.com/jori-nordic/zephyr/blob/a391f85f3deec1370fe98ca1272e1169b0a03fc4/subsys/bluetooth/host/att.c#L720

This `rsp_buf` is only ever read/used by `bt_att_chan_create_pdu`. And when it is, we take a reference (see above). 

The ref take used to be where you want it but that was problematic: 
Some RSP handlers will try to allocate another buffer for the err rsp if the rsp handling fails. 
Doing so, it first does a first `unref`, so the ref we'd take here is consumed, then it 'allocates' another one, but whatever is in `rsp_buf` ends up being used again. Then before returning from `bt_att_recv`, we'd `unref`, thinking we just unrefed the ref we took at the start -> buffer is destroyed and lower layers freak out.

It's a bit hard to explain, but I can try to reproduce the failure and attach the logs/stacktrace if you're unconvinced.

What I did here is take the reference where the buffer is "produced": if `bt_att_chan_create_pdu` is the one creating the buffer from ATT's point of view. So I don't have to change the semantics of all the att buffer handling. They can allocate and unref at will and it will just work.

This `att_chan->rsp_buf = buf;` line only stores the "option" to steal the REQ buffer. If it ends up stolen, then whatever steals it `bt_att_chan_create_pdu` in this case will take the ref.

jori-nordic replied:
that said, I can probably take another ref just in case, that shouldn't be a problem

jori-nordic replied:
seems like I remembered wrong, the tests pass with the extra ref. Lemme know if that's good enough

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -111,8 +111,12 @@ struct cmd_data {
 static struct cmd_data cmd_data[CONFIG_BT_BUF_CMD_TX_COUNT];
 
+#if defined(CONFIG_BT_CONN)
+static struct acl_data acl_data[CONFIG_BT_BUF_ACL_RX_COUNT];
```
This needs  a fix in `conn_internal.h`:

```
struct acl_data {
	/* Extend the bt_buf user data */
	struct bt_buf_data buf_data;
```

jori-nordic replied:
also moved the struct def since it's only used in hci_core.c now

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3934,6 +3934,16 @@ static void rx_work_handler(struct k_work *work)
+bool bt_is_wq(void)
+{
+#if defined(CONFIG_BT_RECV_WORKQ_BT)
+	return ((k_current_get() == &k_sys_work_q.thread) ||
```
Please add a comment about why this is necessary

jori-nordic replied:
extracted in a separate PR
